### PR TITLE
Bug 1816879 - Part 1: New credit cards autofill related UI tests and other refactoring work

### PR DESCRIPTION
**Pull request by AndiAJ**
Bug 1816879 - Part 1: New credit cards autofill related UI tests and other refactoring work

Summary:

- Moved existing credit card autofill related UI tests to a new CreditCardAutofillTest test class.
- Created new UI tests to improve the "Full functional" test suite coverage:
   `verifyCreditCardsSectionTest` ✅ successfully passed 100x on Firebase
   `verifyManageCreditCardsPromptOptionTest` ✅ successfully passed 100x on Firebase
   `verifyCreditCardsAutofillToggleTest` ✅ successfully passed 100x on Firebase
   `verifyEditCardsViewTest` ✅ successfully passed 100x on Firebase
   `verifyEditedCardIsSavedTest` ✅ successfully passed 100x on Firebase

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1816879

### CODE REVIEW

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/CreditCardAutofillTest.kt [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+        const val MOCK_EXPIRATION_MONTH = "February"
+        val MOCK_EXPIRATION_YEAR = (LocalDate.now().year + 1).toString()
+    }
+
+    object SecondCreditCard {
+        const val MOCK_CREDIT_CARD_NUMBER = "2720994326581252"
+        const val MOCK_LAST_CARD_DIGITS = "1252"
+        const val MOCK_NAME_ON_CARD = "Mastercard"
+        const val MOCK_EXPIRATION_MONTH = "March"
+        val MOCK_EXPIRATION_YEAR = (LocalDate.now().year + 2).toString()
```
Comment (AaronMT) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/CreditCardAutofillTest.kt:33]: Mind using similar mock data in https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/components/feature/prompts/src/test/java/mozilla/components/feature/prompts/creditcard/CreditCardsAdapterTest.kt ? I just want to be careful with credit card mock data.

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/CreditCardAutofillTest.kt [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+import org.mozilla.fenix.helpers.TestAssetHelper
+import org.mozilla.fenix.helpers.TestHelper.exitMenu
+import org.mozilla.fenix.ui.robots.homeScreen
+import org.mozilla.fenix.ui.robots.navigationToolbar
+import java.time.LocalDate
+
+class CreditCardAutofillTest {
+    private lateinit var mockWebServer: MockWebServer
+
+    object FirstCreditCard {
```
Comment (AaronMT) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/CreditCardAutofillTest.kt:20]: Maybe `MockCreditCard1` and `MockCreditCard2`

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/SettingsSubMenuAutofillRobot.kt [hunk 3]
```diff
@@ -314,14 +410,22 @@ private val toolbarDeleteAddressButton = itemWithResId("$packageName:id/delete_a
...
 private val cardNumberTextInput = mDevice.findObject(UiSelector().resourceId("$packageName:id/card_number_input"))
 private val nameOnCardTextInput = mDevice.findObject(UiSelector().resourceId("$packageName:id/name_on_card_input"))
 private val expiryMonthDropDown = mDevice.findObject(UiSelector().resourceId("$packageName:id/expiry_month_drop_down"))
 private val expiryYearDropDown = mDevice.findObject(UiSelector().resourceId("$packageName:id/expiry_year_drop_down"))
 private val savedCreditCardNumber = mDevice.findObject(UiSelector().resourceId("$packageName:id/credit_card_logo"))
-private val deleteCreditCardButton = mDevice.findObject(UiSelector().resourceId("$packageName:id/delete_credit_card_button"))
+private val toolbarDeleteCreditCardButton = mDevice.findObject(UiSelector().resourceId("$packageName:id/delete_credit_card_button"))
+private val toolbarSaveCreditCardButton = itemWithResId("$packageName:id/save_credit_card_button")
+private val deleteCardButton = itemContainingText(getStringResource(R.string.credit_cards_delete_card_button))
```
Comment (AaronMT) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/SettingsSubMenuAutofillRobot.kt:428]: Let's just be consistent with having `credit` in our function names.
`deleteCreditCardButton`

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/SettingsSubMenuAutofillRobot.kt [hunk 4]
```diff
@@ -314,14 +410,22 @@ private val toolbarDeleteAddressButton = itemWithResId("$packageName:id/delete_a
 private val cancelDeleteAddressButton = onView(withId(android.R.id.button2)).inRoot(RootMatchers.isDialog())
 private val confirmDeleteAddressButton = onView(withId(android.R.id.button1)).inRoot(RootMatchers.isDialog())
 
+private val creditCardsSectionTitle = itemContainingText(getStringResource(R.string.preferences_credit_cards))
+private val saveAndAutofillCardsOption = itemContainingText(getStringResource(R.string.preferences_credit_cards_save_and_autofill_cards))
+private val saveAndAutofillCardsSummary = itemContainingText(getStringResource(R.string.preferences_credit_cards_save_and_autofill_cards_summary))
+private val syncCardsAcrossDevicesButton = itemContainingText(getStringResource(R.string.preferences_credit_cards_sync_cards_across_devices))
```
Comment (AaronMT) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/SettingsSubMenuAutofillRobot.kt:416]: I would include `credit` in the function names for consistency.

**Review by AaronMT — CHANGES_REQUESTED** [review]
Looks good to me, couple renames and data change.

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/CreditCardAutofillTest.kt (hunk 2 above)
Comment (AndiAJ) [reply fenix/app/src/androidTest/java/org/mozilla/fenix/ui/CreditCardAutofillTest.kt]: Done !

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/SettingsSubMenuAutofillRobot.kt (hunk 3 above)
Comment (AndiAJ) [reply fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/SettingsSubMenuAutofillRobot.kt]: Done !

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/SettingsSubMenuAutofillRobot.kt (hunk 4 above)
Comment (AndiAJ) [reply fenix/app/src/androidTest/java/org/mozilla/fenix/ui/robots/SettingsSubMenuAutofillRobot.kt]: Done !

File: fenix/app/src/androidTest/java/org/mozilla/fenix/ui/CreditCardAutofillTest.kt (hunk 1 above)
Comment (AndiAJ) [reply fenix/app/src/androidTest/java/org/mozilla/fenix/ui/CreditCardAutofillTest.kt:33]: I left the credentials as they were. (I couldn't save the credit card with the credentials from the above link)

When the feature was **first** tested, my fellow colleagues added [pairing of credentials](https://testrail.stage.mozaws.net/index.php?/suites/view/3192&group_by=cases:section_id&group_id=227739&group_order=asc) that we can use.

image


Considering the fact that these are also used when performing manual tests, I would say it's safe to use them.

When I created the first "MockCreditCard" a while ago, I've used the same approach. ( used one of the pairings from our manual tests )

**Comment by AndiAJ** [comment]
Thanks for the suggestions @AaronMT  !

I've included `credit` everywhere in the function names as well as in variable names.

Left a comment regarding the credentials for the second mock credit card.

**Review by AaronMT — APPROVED** [review]

**Pull request merged**